proto/ptypes/interceptor.go

```diff
@@ -0,0 +1,74 @@
+package ptypes
+
+import (
+	context "context"
+
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+func WithIdentityValidationControllerServerInterceptor(volumeName, instanceName string) grpc.ServerOption {
+	return grpc.UnaryInterceptor(identityValidationServerInterceptor(volumeName, instanceName, "controller"))
+}
+
+func WithIdentityValidationReplicaServerInterceptor(volumeName, instanceName string) grpc.ServerOption {
+	return grpc.UnaryInterceptor(identityValidationServerInterceptor(volumeName, instanceName, "replica"))
+}
+
+func identityValidationServerInterceptor(volumeName, instanceName, serverType string) grpc.UnaryServerInterceptor {
+	// Use a closure to remember the correct volumeName and/or instanceName.
+	return func(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+		md, ok := metadata.FromIncomingContext(ctx)
+		if ok {
+			incomingVolumeName, ok := md["volume-name"]
+			// Only refuse to serve if both client and server provide validation information.
+			if ok && volumeName != "" {
+				log := logrus.WithFields(logrus.Fields{"method": info.FullMethod,
+					"clientVolumeName": incomingVolumeName[0], "serverVolumeName": volumeName})
+				if incomingVolumeName[0] != volumeName {
```

Comment:
There is no practical way this can happen, as all "correct" methods of generating the metadata on the client side (including the one we use) would at least provide `incomingVolumeName[0] == ""`. However, I did experiment with generating the metadata "incorrectly" and was able to trigger a panic (https://go.dev/play/p/ybP0IwGjvrX), so I am fixing this along with your comment above.